#### 파이썬은 snake_case와 CamelCase를 모두 사용
- itemlist : item_list itemList
- loginstatus : login_status loginStatus
- 캐멀케이스(대문자로 시작)클래스
- 스네이크 케이스(소문자로 시작) 뒤에()가 있다 - 함수
- 스네이크 케이스(소문자로 시작) 뒤에()가 없다 - 변수

In [ ]:
# 연산자
a = 10
b = 5
c = 20

print(a+b+c)
print(a//b) #몫
print(b%c) #나머지
print(a/c)

35
2
5
0.5


In [ ]:
a = 255
b = 300

if a >b :
    print(a)
else :
    print(b)
print("end")

300
end


#### 자료형
- 자료형 또는 데이터 타입이란 숫자, 문자 등과 같이 여러 종류의 데이터를 구분하기 위한 분류
- 파이썬의 자료형은 크게 숫자(numbers), 시퀀스(sequence), 매핑(mapping) 등으로 나눌 수 있다.
- 파이썬의 기본 자료형
  - 수치형
    - 정수형 : int는 정수(integer)를 나타낸다. 양의 정수와 음의 정수, 숫자 0
    - 실수형 : float는 원래 부동소수점수(floating-point number)를 가리키는데, 지금은 단순히 소수점 이하를 표현할 수 있는 수이다.
    - 복소수형 : 복소수를 complex로 나타내고 제곱하면 -1이 되는 수 i를 ‘허수(imaginary number)’라고 하는데 허수 i를 j로 표현
  - 시퀀스 : 문자열(str), 리스트(list), 튜플(tuple), 사용자 정의 클래스가 시퀀스에 속한다. for 문에서 사용할 수 있는 것들이 바로 시퀀스
    - 문자열 : 문자를 한 줄로 표현하며 문자열 인덱스를 이용해 문자열의 일부를 복사
    - 리스트 : 대괄호([ ])로 감싸 주고 각 요솟값은 쉼표(,)로 구분
    - 튜플 : 튜플은 ( )으로 둘러싸고 각 요솟값은 쉼표(,)로 구분
  - 매핑
    - 사전 : 딕셔너리(dict)는 키(key)와 값(value)의 짝으로 이뤄지는데 이런 것을 매핑
  - 집합 : 집합을 표현하는 세트(set)
  - 블린 : 참, 거짓을 표현

In [ ]:
# 정수형 실수형
i1 = 3
f1 = 3.4
print(i1, type(i1))
print(f1, type(f1))

3 <class 'int'>
3.4 <class 'float'>


In [ ]:
# 정수(int)
print(int(True))
print(int(False))
print(int('10000'))
print(int(3.14))
print(int(float('3.33')))

1
0
10000
3
3


In [ ]:
print(int(True)+100)
print(int(True+1))
print(int(False+True))
print(float(False+True+True+3))
print(float('3.3')+3.3)


101
2
1
5.0
6.6


In [ ]:
a= 10
b=2.3
print(a+b)
print(a*b)
print(a/b)
print(a//b)
print(a%b)
print(a**b)
print(2**3)

12.3
23.0
4.347826086956522
4.0
0.8000000000000007
199.52623149688787
8


In [ ]:
%%time
a = 100
b = 50
print(a+b)
c = '400'
d = '22'
print(c+d)
print(a+int(c))
print(float(b)+int(d))
print(int(c)+int(d))
print(str(a)+c)

150
40022
500
72.0
422
100400
CPU times: user 4.04 ms, sys: 0 ns, total: 4.04 ms
Wall time: 4.05 ms


CPU times: user 4.04 ms, sys: 0 ns, total: 4.04 ms
컴퓨터가 인지하는 시간
Wall time: 4.05 ms (컴퓨터가 작업을 처리하는 시간에 대해 사람이 인지하는 시간)

In [ ]:
# 사용자입력
a = int(input("수를 입력하세요 1 : "))
b = int(input("수를 입력하세요 2 : "))

def add(a,b):
    return a+b

print(add(a,b))

수를 입력하세요 1 : 20
수를 입력하세요 2 : 340
360


In [ ]:
num1 = int(input("첫번째 숫자 : "))
num2 = int(input("두번재 숫자 : "))

def add_(num1,num2):
    result = num1+num2
    return result

print(add_(num1,num2))

첫번째 숫자 : 25
두번재 숫자 : 45
70


In [ ]:
# 실수+실수 -> 실수
num1 = 3
num2 = 3.3

print(float(num1)+float(num2))
print(float(num1)+num2)


6.3
6.3


In [ ]:
# 정수+실수 = 실수 #실수가 가장 큰 범위이므로
i1 = int(input("정수를 입력하세요"))
f1 = float(input("실수를 입력하세요"))

print(i1+f1)

정수를 입력하세요2
실수를 입력하세요3.3
5.3


In [ ]:
a=52
b=23.273
print(str(a)+str(b))

# 변수에 수정한 값 다시 저장
a= str(a)
b=str(b)
print(a+b)

5223.273
5223.273


In [ ]:
# 문자열 : "",'',''' ''', """ """
# p1 = 'python's value is great'
# 종료 구분 기호가 모호하다

p1 = "python's value is great"
print(p1)
p2 = "'hello'"
print(p2)
p3 = '"hello"'
print(p3)
p4 = "'h'ello"
print(p4)
# 이스케이프 문자 활용 -> 역슬래쉬 \
p5 = "python\'s'"
print(p5)
p6 = "\'안녕\' 반가워"
print(p6)

python's value is great
'hello'
"hello"
'h'ello
python's'
'안녕' 반가워


### 백슬래쉬(\)
- 백슬래쉬(\)는 문자열 안에서 작은따옴표(') 를 문자열의 끝을 의미하지 않는 일반 문자로 포함시키기 위해 사용
- 이 기술을 이스케이프라고함

In [ ]:
# \n 행바꿈
y1 = "Once you study data analysis \nYou need Python"
print(y1)
y2 = "you want to study \t'python'"
print(y2)

Once you study data analysis 
You need Python
you want to study 	'python'


In [ ]:
y3 = '''
Once you study data analysis
You need Python
'''
print(y3)

y3 = '''
Once you study data analysis\nYou need Python
'''
print(y3)

y3 = '''Once you study data analysis
You need Python\nwhat isss
\t you know
'''
print(y3)

y3 = """Once you study data analysis
You need \'Python\'"""
print(y3)


Once you study data analysis
You need Python


Once you study data analysis
You need Python

Once you study data analysis
You need Python
what isss
	 you know

Once you study data analysis
You need 'Python'


In [ ]:
a1 = 'python'
a2 = ' is easy to learn'
print(a1+a2)
print("!"*23)
print(a1 *3)
print(3**2)
print("str"*2)

python is easy to learn
!!!!!!!!!!!!!!!!!!!!!!!
pythonpythonpython
9
strstr


In [ ]:
# 아래와 같이 출력할 수 있는 string을 만드세요
# "What a wonderful world!"
# he said loudly

s1 = '''"What a wonderful world!"
he said loudly'''

print(s1)

"What a wonderful world!"
he said loudly


In [ ]:
s1 = '''
"What a wonderful world!"
he said loudly
'''
print(s1.strip())
# strip() 양쪽 공백이나 원하는 문자 제거

"What a wonderful world!"
he said loudly


In [ ]:
# 다른 타입으 숫자 2개를 입력받아 큰 수를 출력하세요
num1 = int(input("정수를 입력하세요 : "))
num2 = float(input("실수를 입력하세요 : "))

if num1>num2:
    print(num1)
else: print(num2)

정수를 입력하세요 : 4
실수를 입력하세요 : 2.2
4


In [ ]:
# 다른 타입으 숫자 2개를 입력받아 큰 수를 출력하세요
num1 = int(input("정수를 입력하세요 : "))
num2 = float(input("실수를 입력하세요 : "))

if num1>num2:
    print(f"{num1}이 큰수입니다")
elif num1<num2:
     print(f"{num2}이 큰수입니다")
else :
    print("같은수입니다")
print("출력이 완료됐습ㄴ다")

정수를 입력하세요 : 5
실수를 입력하세요 : 2
5이 큰수입니다
출력이 완료됐습ㄴ다


In [ ]:
# 자료형 List
# 리스트는 []로 푯하며 []안의 요소를 콤마로 구분하여 순서있게 나열
list1 = [1,2,3,4,5]
list2 = ['a','b','c']
list3 = [1,'a',(1,2,3,4,5),['a','b','c'],True]
print(list1)
print(list2)
print(list3,type(list3))
print(list3[2],type(list3[2]))
print(list3[-1],type(list3[-1]))

[1, 2, 3, 4, 5]
['a', 'b', 'c']
[1, 'a', (1, 2, 3, 4, 5), ['a', 'b', 'c'], True] <class 'list'>
(1, 2, 3, 4, 5) <class 'tuple'>
True <class 'bool'>


In [ ]:
# 인덱싱
li1 = [1,2,3,4,5]
print(li1[-1])
print(li1[0:3])

5
[1, 2, 3]


In [ ]:
li2 = [1,'a',(1,2,3,4,5),['a','b','c'],True]

print(li2[3][2])
print(li2[2][2])
print(li2[3][0])

c
3
a


In [ ]:
# 리스트 수정, 삭제
li3 = [0,1,2,3,4,5,6,8]
li4 = ['a','b','c']
li3[-1] = 7
print(li3)

li4[-1]= 'd'
print(li4)

del li4[1]
print(li4)

del li3[0]
print(li3)

[0, 1, 2, 3, 4, 5, 6, 7]
['a', 'b', 'd']
['a', 'd']
[1, 2, 3, 4, 5, 6, 7]


In [ ]:
# 끝에서부터 인덱싱 -1
a = [1,2,3,4,5,6,7,8,9]
print(a[-5:-1:2])
print(a[-1:-3:-1])
print(a[-3])

[5, 7]
[9, 8]
7


In [ ]:
# 리스트 확장
h = [1,2,3]
print(h)
h.extend([4,5,6,7]) # 반복 가능한 객체의 모든 요소를 리스트에 추가
print(h)
h.insert(0,0)
print(h)
h.append(8) #맨뒤에 객체 추가(단일만 가능)
print(h)
h.extend([7]) #하나도 가능하나 []로 묶어야함
print(h)
h.extend([True,"str"])
print(h)
print(h[-1],type(h[-1]))
print(h[-2],type(h[-2]))

[1, 2, 3]
[1, 2, 3, 4, 5, 6, 7]
[0, 1, 2, 3, 4, 5, 6, 7]
[0, 1, 2, 3, 4, 5, 6, 7, 8]
[0, 1, 2, 3, 4, 5, 6, 7, 8, 7]
[0, 1, 2, 3, 4, 5, 6, 7, 8, 7, True, 'str']
str <class 'str'>
True <class 'bool'>


Q. list1에서 다음을 수행하세요
- 33을 출력
- 82를 리스트에 추가
- 87의 인덱스 구하기 - 도전
- 인덱스 3에서 10까지의 값을 출력하고 list2에 저장한 후 내림차순 정렬하기 - 도전
- 39를 11로 변경
- [69,45,58] 출력
- 짝수 인덱스의 값으로 구성된 리스트 출력하기
- 인덱스가 가장 큰수를 삭제하기
- 인덱스 3, 5인 값으로 4칙 연산하기

list1 = [58, 45, 69, 19,  4, 87, 29, 13, 39, 15, 54, [8, 33, 11], 27, 49, 63, 98, 22, 82]

In [66]:
list1 = [58, 45, 69, 19,  4, 87, 29, 13, 39, 15, 54, [8, 33, 11], 27, 49, 63, 98, 22, 82]

print(list1[-7][1])

list1.append(82)
print(list1)

print(list1.index(87))
# 변수명.index(구하는값,시작위치,끝위치)

list2 = list1[3:11]
print(list2)
list2_sort = sorted(list2, reverse=True)
print(list2_sort)

list1[8] = 11
print(list1[8])

print(list1[2::-1])

print(list1[::2])

del list1[-1]
print(list1)

print(list1[3]+list1[5])
print(list1[3]-list1[5])
print(list1[3]*list1[5])
print(list1[3]/list1[5])

33
[58, 45, 69, 19, 4, 87, 29, 13, 39, 15, 54, [8, 33, 11], 27, 49, 63, 98, 22, 82, 82]
5
[19, 4, 87, 29, 13, 39, 15, 54]
[87, 54, 39, 29, 19, 15, 13, 4]
11
[69, 45, 58]
[58, 69, 4, 29, 11, 54, 27, 63, 22, 82]
[58, 45, 69, 19, 4, 87, 29, 13, 11, 15, 54, [8, 33, 11], 27, 49, 63, 98, 22, 82]
106
-68
1653
0.21839080459770116


#### sort 와 sorted 차이점

1. 메서드 vs 함수:
- sort(): 리스트 객체의 내장 메서드입니다. 즉, 리스트에서만 사용할 수 있습니다.
- sorted(): 내장 함수로, 어떤 반복 가능한(iterable) 객체에도 사용될 수 있습니다. 예를 들면 리스트, 튜플, 딕셔너리, 문자열 등에 사용할 수 있습니다.

2. 반환 값:
- sort(): 리스트를 원 위치에서(in-place) 정렬하고 None을 반환합니다. 따라서 원래의 리스트 자체가 변경됩니다. -> 반환하는 값이 없다
- sorted(): 정렬된 새로운 리스트를 반환합니다. 원래의 객체는 변경되지 않습니다.

3. 유용성:
- sort(): 리스트에서만 작동하기 때문에 리스트만 정렬할 수 있습니다.
- sorted(): 다양한 객체를 정렬할 수 있으며 결과는 항상 리스트로 반환됩니다.

In [ ]:
# sort()
my_list = [3,1,2]
my_list.sort()
print(my_list)

# sorted()
my_tuple = (3,1,2)
new_list = sorted(my_tuple)
print(my_tuple)
print(my_list)

# sort는 원래의 리스트가 변환되지만 sorted는 원래의 객체는 변경되지 않는다

[1, 2, 3]
(3, 1, 2)
[1, 2, 3]


In [ ]:
li1 = [5,6,9,7,8]
new_list = li1.sort()
print(li1)
print(new_list)
# sort()는 리스트 객체의 내장 메소드로 원래의 리스트가 변경된다 그러므로 새 변수에 저장을 해도 값이 없다

[5, 6, 7, 8, 9]
None


In [ ]:
li2 = [5,6,9,8,7]
new_list1 = sorted(li2)
print(li2)
print(new_list1)
# sorted는 기존 객체에는 변화가 없고 정렬된 새로운 객체를 만들어낸다

[5, 6, 9, 8, 7]
[5, 6, 7, 8, 9]


In [ ]:
list1 = [58, 45, 69, 19,  4, 87, 29, 13, 39, 15, 54, [8, 33, 11], 27, 49, 63, 98, 22, 82]

list2 = list1[3:11]
l_sort = sorted(list2,reverse=True)
print(l_sort)

[87, 54, 39, 29, 19, 15, 13, 4]


In [ ]:
list3 = list1[3:11]
list3.sort(reverse=True)
print(list3)

[87, 54, 39, 29, 19, 15, 13, 4]


In [ ]:
# Task1_0425. 주어진 숫자 리스트에서 최소값과 최대값을 찾아 출력하세요.

numbers = [58, 45, 69, 19, 4, 87, 29, 13, 39, 15]
print(max(numbers))
print(min(numbers))

87
4


In [ ]:
# Task2_0425. 주어진 숫자 리스트의 모든 요소의 합계와 평균을 계산하고 출력하세요

numbers = [58, 45, 69, 19, 4, 87, 29, 13, 39, 15]
total = sum(numbers)
print(total)
print(total/len(numbers))
ave = sum(numbers)/len(numbers)
print(ave)

378
37.8
37.8


In [2]:
# Task3_0425. 주어진 리스트에서 특정 요소가 등장하는 모든 인덱스를 리스트로 만들어 출력하세요.

items = ['apple', 'banana', 'cherry', 'apple', 'cherry', 'apple']
# target = 'apple'

index_items = []
for i in items :
    if i == "apple":
        index_items.append(i)
print(index_items)

# 출력 [0,3,5]


['apple', 'apple', 'apple']


In [42]:
ind = 0
items_list = []
for i in range(len(items)):
    if items.index('apple',ind):
        ind2 = items.index('apple',ind)
        items_list.append(ind2)
    ind = ind+1
print(items_list)

[3, 3, 3, 5, 5]


In [53]:
index2 = [index for index, item in enumerate(items) if item == 'apple']

print(index2)
# enumerate 인덱스 값 형식으로 추출해주는 함수

[0, 3, 5]


In [52]:
# Task4_0425. 주어진 리스트에서 연속해서 반복되는 요소만 제거하고, 결과 리스트를 반환하세요. 단, 처음 등장하는 요소는 유지해야 합니다.

# 예를 들어, ['a', 'a', 'b', 'c', 'c', 'c', 'd', 'e', 'e']가 입력되면, ['a', 'b', 'c', 'd', 'e']를 출력해야 합니다.
data = ['a', 'a', 'b', 'c', 'c', 'c', 'd', 'e', 'e']

from collections import Counter
count_data = Counter(data)
print(count_data)


index2 = [index for index, item in count_data.items() if item > 1 or item == 1]
print(index2)



# 그 값이 1개 초과라면
# 가진 값에서 -1개를 해서 추출

Counter({'c': 3, 'a': 2, 'e': 2, 'b': 1, 'd': 1})
['a', 'b', 'c', 'd', 'e']


In [ ]:
data = ['a', 'a', 'b', 'c', 'c', 'c', 'd', 'e', 'e']

# set은 중복값을 허용하지 않으므로
data1 = set(data)
data = list(data1)
print(data)
data1 = sorted(data)
print(data1)

['c', 'a', 'b', 'd', 'e']
['a', 'b', 'c', 'd', 'e']


In [ ]:
# Task5_0425 주어진 정수 리스트와 회전 횟수 k에 대해 리스트를 오른쪽으로 k만큼 회전시킨 결과를 반환하세요.
# k가 리스트의 길이보다 클 수 있으며, 이 경우 k를 리스트 길이로 나눈 나머지만큼 실제 회전시키면 됩니다.

# 예를 들어, [1, 2, 3, 4, 5]와 k=2가 주어지면, 결과는 [4, 5, 1, 2, 3]이 되어야 합니다.
li1 = [1,2,3,4,5]
# n > k / len(li1) % k 만큼 회전
# k만큼 뒤에 있는 수가 앞으로 ...